src/replica/duplication/replica_follower.h

```diff
@@ -41,15 +41,41 @@ class replica_follower : replica_base
 private:
     replica *_replica;
     task_tracker _tracker;
+    bool _duplicating_checkpoint{false};
+    mutable zlock _lock;
 
     std::string _master_cluster_name;
     std::string _master_app_name;
     std::vector<rpc_address> _master_meta_list;
+    partition_configuration _master_replica_config;
 
     bool need_duplicate{false};
 
     void init_master_info();
 
+    std::string master_replica_name()
+    {
+        std::string app_info = fmt::format("{}.{}", _master_cluster_name, _master_app_name);
+        if (_master_replica_config.primary != rpc_address::s_invalid_address) {
+            return fmt::format("{}({}|{})",
+                               app_info,
+                               _master_replica_config.primary.to_string(),
+                               _master_replica_config.pid.to_string());
+        }
+        return app_info;
+    }
+
+    void async_duplicate_checkpoint_from_master_replica();
```

Comment:
In your future pull request, I recommend you can add comments in your head files or add some duplication flow graphs.

Reply:
Good! however it has a little complex, add the link https://github.com/apache/incubator-pegasus/issues/892 is ok?

Reply:
Okay, maybe you can finally add a rfc document.